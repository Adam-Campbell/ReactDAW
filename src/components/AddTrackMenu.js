import React, { Component } from 'react';
import { connect } from 'react-redux';
import * as ActionCreators from '../actions';
import { generateId } from '../helpers';
import { synthTypes, UIColors } from '../constants';

class AddTrackMenu extends Component {
    constructor(props) {
        super(props);
        this.state = {
            menuIsOpen: false
        };
    }

    openMenu = () => {
        this.setState({
            menuIsOpen: true
        });
    }

    closeMenu = () => {
        this.setState({
            menuIsOpen: false
        });
    }

    handleAddChannel = (instrumentType) => {
        this.props.addChannel(
            generateId(),
            `Channel ${this.props.numberOfChannels}`,
            UIColors.pink,
            generateId(),
            instrumentType
        );
        this.closeMenu();
    }


    render() {
        const { menuIsOpen } = this.state;
        return (
            <div className="add-track-menu">
                <button 
                    className="add-track-menu__button"
                    onClick={this.openMenu}
                >+ Add track</button>
                <div 
                    className={`add-track-menu__item-container ${menuIsOpen ? 'show-menu' : ''}`}
                >
                    <ul className="add-track-menu__list">
                        <li 
                            className="add-track-menu__menu-item"
                            onClick={() => this.handleAddChannel(synthTypes.default)}
                        >Synth</li>
                        <li 
                            className="add-track-menu__menu-item"
                            onClick={() => this.handleAddChannel(synthTypes.am)}
                        >AM Synth</li>
                        <li 
                            className="add-track-menu__menu-item"
                            onClick={() => this.handleAddChannel(synthTypes.fm)}
                        >FM Synth</li>
                        <li 
                            className="add-track-menu__menu-item"
                            onClick={() => this.handleAddChannel(synthTypes.duo)}
                        >Duo Synth</li>
                        <li 
                            className="add-track-menu__menu-item"
                            onClick={() => this.handleAddChannel(synthTypes.mono)}
                        >Mono Synth</li>
                        <li 
                            className="add-track-menu__menu-item"
                            onClick={this.closeMenu}
                        >X</li>
                    </ul> 
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    numberOfChannels: state.channels.length + 1
});

export default connect(
    mapStateToProps,
    {
        addChannel: ActionCreators.addChannel
    }
)(AddTrackMenu);