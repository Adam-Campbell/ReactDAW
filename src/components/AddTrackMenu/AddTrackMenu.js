import React from 'react';
import PropTypes from 'prop-types';
import { synthTypes } from '../../constants';

const AddTrackMenu = props => (
    <div className="add-track-menu">
        <button 
            className="add-track-menu__button"
            onClick={props.openMenu}
        >+ Add track</button>
        <div 
            className={`add-track-menu__item-container ${props.menuIsOpen ? 'show-menu' : ''}`}
        >
            <ul className="add-track-menu__list">
                <li 
                    className="add-track-menu__menu-item"
                    onClick={() => props.handleAddChannel(synthTypes.default)}
                >Synth</li>
                <li 
                    className="add-track-menu__menu-item"
                    onClick={() => props.handleAddChannel(synthTypes.am)}
                >AM Synth</li>
                <li 
                    className="add-track-menu__menu-item"
                    onClick={() => props.handleAddChannel(synthTypes.fm)}
                >FM Synth</li>
                <li 
                    className="add-track-menu__menu-item"
                    onClick={() => props.handleAddChannel(synthTypes.duo)}
                >Duo Synth</li>
                <li 
                    className="add-track-menu__menu-item"
                    onClick={() => props.handleAddChannel(synthTypes.mono)}
                >Mono Synth</li>
                <li 
                    className="add-track-menu__menu-item"
                    onClick={props.closeMenu}
                >X</li>
            </ul> 
        </div>
    </div>
);

AddTrackMenu.propTypes = {
    menuIsOpen: PropTypes.bool.isRequired,
    openMenu: PropTypes.func.isRequired,
    closeMenu: PropTypes.func.isRequired,
    handleAddChannel: PropTypes.func.isRequired
};

export default AddTrackMenu;