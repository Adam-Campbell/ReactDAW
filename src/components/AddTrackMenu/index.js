import React, { Component } from 'react';
import { connect } from 'react-redux';
import * as ActionCreators from '../../actions';
import { generateId } from '../../sharedUtils';
import { UIColors } from '../../constants';
import AddTrackMenu from './AddTrackMenu';

export class AddTrackMenuContainer extends Component {
    constructor(props) {
        super(props);
        this.state = {
            menuIsOpen: false
        };
    }

    openMenu = () => {
        this.setState({
            menuIsOpen: true
        });
    }

    closeMenu = () => {
        this.setState({
            menuIsOpen: false
        });
    }

    handleAddChannel = (instrumentType) => {
        this.props.addChannel(
            generateId(),
            `Channel ${this.props.numberOfChannels + 1}`,
            UIColors.pink,
            generateId(),
            instrumentType
        );
        this.closeMenu();
    }

    render() {
        return <AddTrackMenu 
            menuIsOpen={this.state.menuIsOpen}
            openMenu={this.openMenu}
            closeMenu={this.closeMenu}
            handleAddChannel={this.handleAddChannel}
        />
    }
}

const mapStateToProps = (state) => ({
    numberOfChannels: state.channels.length
});

export default connect(
    mapStateToProps,
    {
        addChannel: ActionCreators.addChannel
    }
)(AddTrackMenuContainer);