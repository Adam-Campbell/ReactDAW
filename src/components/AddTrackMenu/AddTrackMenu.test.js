import React from 'react';
import { shallow } from 'enzyme';
import { AddTrackMenuContainer } from './';
import AddTrackMenu from './AddTrackMenu';

const mockedAddChannel = jest.fn();

test('container component renders correctly', () => {
    const component = shallow(
        <AddTrackMenuContainer 
            numberOfChannels={0}
            addChannel={mockedAddChannel}
        />
    );
    expect(component).toMatchSnapshot();
    expect(component.find(AddTrackMenu)).toHaveLength(1);
});

test('presentational component renders correctly', () => {
    const component = shallow(
        <AddTrackMenu 
            menuIsOpen={true}
            openMenu={jest.fn()}
            closeMenu={jest.fn()}
            handleAddChannel={jest.fn()}
        />
    );
    expect(component).toMatchSnapshot();
});

test('component handles switching between open and closed states correctly', () => {
    const component = shallow(
        <AddTrackMenuContainer 
            numberOfChannels={0}
            addChannel={mockedAddChannel}
        /> 
    );
    expect(component.find(AddTrackMenu).props().menuIsOpen).toEqual(false);
    component.instance().openMenu();
    expect(component.find(AddTrackMenu).props().menuIsOpen).toEqual(true);
    component.instance().closeMenu();
    expect(component.find(AddTrackMenu).props().menuIsOpen).toEqual(false);
});

test('component calls the addTrack handler when a list item is clicked', () => {
    const component = shallow(
        <AddTrackMenuContainer 
            numberOfChannels={2}
            addChannel={mockedAddChannel}
        /> 
    );
    const presentational = component.first().shallow();
    expect(mockedAddChannel).not.toHaveBeenCalled();
    presentational.find('li.add-track-menu__menu-item').first().simulate('click');
    expect(mockedAddChannel).toHaveBeenCalledTimes(1);
    expect(mockedAddChannel.mock.calls[mockedAddChannel.mock.calls.length-1][1]).toBe('Channel 3');
});