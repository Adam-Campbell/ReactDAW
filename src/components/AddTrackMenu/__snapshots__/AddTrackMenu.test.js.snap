// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`container component renders correctly 1`] = `
<AddTrackMenu
  closeMenu={[Function]}
  handleAddChannel={[Function]}
  menuIsOpen={false}
  openMenu={[Function]}
/>
`;

exports[`presentational component renders correctly 1`] = `
<div
  className="add-track-menu"
>
  <button
    className="add-track-menu__button"
    onClick={[MockFunction]}
  >
    + Add track
  </button>
  <div
    className="add-track-menu__item-container show-menu"
  >
    <ul
      className="add-track-menu__list"
    >
      <li
        className="add-track-menu__menu-item"
        onClick={[Function]}
      >
        Synth
      </li>
      <li
        className="add-track-menu__menu-item"
        onClick={[Function]}
      >
        AM Synth
      </li>
      <li
        className="add-track-menu__menu-item"
        onClick={[Function]}
      >
        FM Synth
      </li>
      <li
        className="add-track-menu__menu-item"
        onClick={[Function]}
      >
        Duo Synth
      </li>
      <li
        className="add-track-menu__menu-item"
        onClick={[Function]}
      >
        Mono Synth
      </li>
      <li
        className="add-track-menu__menu-item"
        onClick={[MockFunction]}
      >
        X
      </li>
    </ul>
  </div>
</div>
`;
